src/functional/iterator/backends/gtfn/gtfn_ir.py

```diff
@@ -0,0 +1,124 @@
+import enum
+from typing import List, Union
+
+from eve import Node
+from eve.traits import SymbolName, SymbolTableTrait
+from eve.type_definitions import StrEnum, SymbolRef
+from functional.iterator.util.sym_validation import validate_symbol_refs
+
+
+@enum.unique
+class GridType(StrEnum):
+    Cartesian = "Cartesian"
+    Unstructured = "unstructured"
+
+
+class Sym(Node):  # helper
+    id: SymbolName  # noqa: A003
+
+
+class Expr(Node):
+    ...
+
+
+class UnaryExpr(Expr):
+    op: str
+    expr: Expr
+
+
+class BinaryExpr(Expr):
+    op: str
+    lhs: Expr
+    rhs: Expr
+
+
+class TernaryExpr(Expr):
+    cond: Expr
+    true_expr: Expr
+    false_expr: Expr
+
+
+class BoolLiteral(Expr):
+    value: bool
+
+
+class IntLiteral(Expr):
+    value: int
+
+
+class FloatLiteral(Expr):
+    value: float  # TODO other float types
+
+
+class StringLiteral(Expr):
+    value: str
+
+
+class OffsetLiteral(Expr):
+    value: Union[int, str]
+
+
+class SymRef(Expr):
+    id: SymbolRef  # noqa: A003
+
+
+class Lambda(Expr, SymbolTableTrait):
+    params: List[Sym]
+    expr: Expr
+
+
+class FunCall(Expr):
+    fun: Expr  # VType[Callable]
+    args: List[Expr]
+
+
+class TemplatedFunCall(Expr):
+    fun: Expr  # VType[Callable]
+    template_args: List[Expr]
+    args: List[Expr]
+
+
+class FunctionDefinition(Node, SymbolTableTrait):
+    id: SymbolName  # noqa: A003
+    params: List[Sym]
+    expr: Expr
+
+    def __eq__(self, other):
+        return isinstance(other, FunctionDefinition) and self.id == other.id
```

Comment:
Why do we have that (also in the IR)? I think we should remove it, the `==` is used in the tests to compare trees, so this basically invalidates all tests including a `FunctionDefinition`.

Reply:
I guess that is from before nodes go default comparison

Reply:
So I have a tiny review work for you: https://github.com/GridTools/gt4py/pull/743